packages/backend/src/functions/ceremony.ts

```diff
@@ -55,3 +75,262 @@ export const stopCeremony = functions.pubsub.schedule(`every 30 minutes`).onRun(
+            penalty: ceremonyInputData.penalty,
+            timeoutType: ceremonyInputData.timeoutMechanismType,
+            coordinatorId: userId,
+            lastUpdated: getCurrentServerTimestampInMillis()
+        })
+
+        // Create a new circuit document (circuits ceremony document sub-collection).
+        for (const circuit of circuits) {
+            // Get a new circuit document.
+            const circuitDoc = await firestore.collection(getCircuitsCollectionPath(ceremonyDoc.ref.id)).doc().get()
+
+            // Prepare tx to write circuit data.
+            batch.create(circuitDoc.ref, {
+                ...circuit,
+                lastUpdated: getCurrentServerTimestampInMillis()
+            })
+        }
+
+        // Send txs in a batch (to avoid race conditions).
+        await batch.commit()
+
+        printLog(`Setup completed for ceremony ${ceremonyDoc.id}`, LogLevel.DEBUG)
+
+        return ceremonyDoc.id
+    }
+)
+
+/**
+ * Prepare all the necessary information needed for initializing the waiting queue of a circuit.
+ * @dev this function will add a new field `waitingQueue` in the newly created circuit document.
+ */
+export const initEmptyWaitingQueueForCircuit = functions.firestore
+    .document(
+        `/${commonTerms.collections.ceremonies.name}/{ceremony}/${commonTerms.collections.circuits.name}/{circuit}`
+    )
+    .onCreate(async (doc: QueryDocumentSnapshot) => {
+        // Prepare Firestore DB.
+        const firestore = admin.firestore()
+
+        // Get circuit document identifier and data.
+        const circuitId = doc.id
+        // Get parent ceremony collection path.
+        const parentCollectionPath = doc.ref.parent.path // == /ceremonies/{ceremony}/circuits/.
+
+        // Define an empty waiting queue.
+        const emptyWaitingQueue: CircuitWaitingQueue = {
+            contributors: [],
+            currentContributor: "",
+            completedContributions: 0,
+            failedContributions: 0
+        }
+
+        // Update the circuit document.
+        await firestore.collection(parentCollectionPath).doc(circuitId).set(
+            {
+                waitingQueue: emptyWaitingQueue,
+                lastUpdated: getCurrentServerTimestampInMillis()
+            },
+            { merge: true }
+        )
+
+        printLog(
+            `An empty waiting queue has been successfully initialized for circuit ${circuitId} which belongs to ceremony ${doc.id}`,
+            LogLevel.DEBUG
+        )
+    })
+
+/// @todo needs refactoring below.
+
+/**
+ * Add Verifier smart contract and verification key files metadata to the last final contribution for verifiability/integrity of the ceremony.
+ */
+export const finalizeLastContribution = functions.https.onCall(
+    async (data: any, context: functions.https.CallableContext): Promise<any> => {
+        if (!context.auth || !context.auth.token.coordinator)
+            printLog(COMMON_ERRORS.GENERR_NO_COORDINATOR, LogLevel.ERROR)
```

Comment:
I see in other functions you use directly:

`logAndThrowError(COMMON_ERRORS.CM_NOT_COORDINATOR_ROLE)`

Reply:
Yep, this is going to be updated when refactoring the finalize command!